golang/src/user-sessions/main.go

```diff
@@ -0,0 +1,393 @@
+}
+
+// UserSessionRecord represents a user session stored in the database
+type UserSessionRecord struct {
+	ID      string
+	Session string
+	UserID  string
+}
+
+// convert makes sure that the JSON has the correct format. "wrap" tells convert
+// whether to wrap the object in a map with "session" as the key.
+func convert(record *UserSessionRecord, wrap bool) (map[string]interface{}, error) {
+	var values map[string]interface{}
+	if record.Session != "" {
+		if err := json.Unmarshal([]byte(record.Session), &values); err != nil {
+			return nil, err
+		}
+	}
+	// We don't want the return value wrapped in a session object, so unwrap it
+	// if it is wrapped.
+	if !wrap {
+		if _, ok := values["session"]; ok {
+			return values["session"].(map[string]interface{}), nil
+		}
+		return values, nil
+	}
+	// We do want the return value wrapped in a session object, so wrap it if it
+	// isn't already.
+	if _, ok := values["session"]; !ok {
+		newmap := make(map[string]interface{})
+		newmap["session"] = values
+		return newmap, nil
+	}
+	return values, nil
+}
+
+// UserSessionsApp is an implementation of the App interface created to manage
+// user sessions.
+type UserSessionsApp struct {
+	db *sql.DB
+}
+
+// New returns a new *UserSessionsApp
+func New(db *sql.DB) *UserSessionsApp {
+	return &UserSessionsApp{
+		db: db,
+	}
+}
+
+// hasSessions returns whether or not the given user has a session already.
+func (u *UserSessionsApp) hasSessions(username string) (bool, error) {
+	query := `SELECT COUNT(s.*)
+              FROM user_sessions s,
+                   users u
+             WHERE s.user_id = u.id
+               AND u.username = $1`
+	var count int64
+	if err := u.db.QueryRow(query, username).Scan(&count); err != nil {
+		return false, err
+	}
+	return count > 0, nil
+}
+
+// getSessions returns a []UserSessionRecord of all of the sessions associated
+// with the provided username.
+func (u *UserSessionsApp) getSessions(username string) ([]UserSessionRecord, error) {
+	query := `SELECT s.id AS id,
+                   s.user_id AS user_id,
+                   s.session AS session
+              FROM user_sessions s,
+                   users u
+             WHERE s.user_id = u.id
+               AND u.username = $1`
+	rows, err := u.db.Query(query, username)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	var sessions []UserSessionRecord
+	for rows.Next() {
+		var session UserSessionRecord
+		if err := rows.Scan(&session.ID, &session.UserID, &session.Session); err != nil {
+			return nil, err
+		}
+		sessions = append(sessions, session)
+	}
+	if err := rows.Err(); err != nil {
+		return sessions, err
+	}
+	return sessions, nil
+}
+
+// insertSession adds a new session to the database for the user.
+func (u *UserSessionsApp) insertSession(username, session string) error {
+	query := `INSERT INTO user_sessions (user_id, session)
+                 VALUES ($1, $2)`
+	userID, err := queries.UserID(u.db, username)
```

Comment:
The database pedant in me is mildly uncomfortable with us doing two queries/database roundtrips here when it could probably be done with one, but combining them would probably require separating the query string from the execution in the `queries` namespace and I'm not sure that's worth it. Just noting it here in case you or someone else has a clever idea that's not too invasive.